File: server/src/main/java/se/kth/id2203/bootstrapping/BootstrapServer.java
/*
 * The MIT License
 *
 * Copyright 2017 Lars Kroll <lkroll@kth.se>.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */
package se.kth.id2203.bootstrapping;

import com.google.common.collect.ImmutableSet;
import java.util.HashSet;
import java.util.Set;
import java.util.UUID;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import se.kth.id2203.bootstrapping.BootstrapServer.State;
import se.kth.id2203.networking.Message;
import se.kth.id2203.networking.NetAddress;
import se.sics.kompics.ClassMatchedHandler;
import se.sics.kompics.ComponentDefinition;
import se.sics.kompics.Handler;
import se.sics.kompics.Negative;
import se.sics.kompics.Positive;
import se.sics.kompics.Start;
import se.sics.kompics.network.Network;
import se.sics.kompics.timer.CancelPeriodicTimeout;
import se.sics.kompics.timer.SchedulePeriodicTimeout;
import se.sics.kompics.timer.Timer;

public class BootstrapServer extends ComponentDefinition {

    final static Logger LOG = LoggerFactory.getLogger(BootstrapServer.class);
    //******* Ports ******
    protected final Negative<Bootstrapping> boot = provides(Bootstrapping.class);
    protected final Positive<Network> net = requires(Network.class);
    protected final Positive<Timer> timer = requires(Timer.class);
    //******* Fields ******
    final NetAddress self = config().getValue("id2203.project.address", NetAddress.class);
    final int bootThreshold = config().getValue("id2203.project.bootThreshold", Integer.class);
    private State state = State.COLLECTING;
    private UUID timeoutId;
    private final Set<NetAddress> active = new HashSet<>();
    private final Set<NetAddress> ready = new HashSet<>();
    private NodeAssignment initialAssignment = null;
    //******* Handlers ******
    protected final Handler<Start> startHandler = new Handler<Start>() {
        @Override
        public void handle(Start e) {
            LOG.info("Starting bootstrap server on {}, waiting for {} nodes...", self, bootThreshold);
            long timeout = (config().getValue("id2203.project.keepAlivePeriod", Long.class) * 2);
            SchedulePeriodicTimeout spt = new SchedulePeriodicTimeout(timeout, timeout);
            spt.setTimeoutEvent(new BSTimeout(spt));
            trigger(spt, timer);
            timeoutId = spt.getTimeoutEvent().getTimeoutId();
            active.add(self);
        }
    };
    protected final Handler<BSTimeout> timeoutHandler = new Handler<BSTimeout>() {
        @Override
        public void handle(BSTimeout e) {
            if (state == State.COLLECTING) {
                LOG.info("{} hosts in active set.", active.size());
                if (active.size() >= bootThreshold) {
                    bootUp();
                }
            } else if (state == State.SEEDING) {
                LOG.info("{} hosts in ready set.", ready.size());
                if (ready.size() >= bootThreshold) {
                    LOG.info("Finished seeding. Bootstrapping complete.");
                    trigger(new Booted(initialAssignment), boot);
                    state = State.DONE;
                }
            } else if (state == State.DONE) {
                suicide();
            }
        }
    };
    protected final Handler<InitialAssignments> assignmentHandler = new Handler<InitialAssignments>() {
        @Override
        public void handle(InitialAssignments e) {
            LOG.info("Seeding assignments...");
            initialAssignment = e.assignment;
            for (NetAddress node : active) {
                trigger(new Message(self, node, new Boot(initialAssignment)), net);
            }
            ready.add(self);
        }
    };
    protected final ClassMatchedHandler<CheckIn, Message> checkinHandler = new ClassMatchedHandler<CheckIn, Message>() {

        @Override
        public void handle(CheckIn content, Message context) {
            active.add(context.getSource());
        }
    };
    protected final ClassMatchedHandler<Ready, Message> readyHandler = new ClassMatchedHandler<Ready, Message>() {
        @Override
        public void handle(Ready content, Message context) {
            ready.add(context.getSource());
        }
    };

    {
        subscribe(startHandler, control);
        subscribe(timeoutHandler, timer);
        subscribe(assignmentHandler, boot);
        subscribe(checkinHandler, net);
        subscribe(readyHandler, net);
    }

    @Override
    public void tearDown() {
        trigger(new CancelPeriodicTimeout(timeoutId), timer);
    }

    private void bootUp() {
        LOG.info("Threshold reached. Generating assignments...");
        state = State.SEEDING;
        trigger(new GetInitialAssignments(ImmutableSet.copyOf(active)), boot);
    }

    static enum State {

        COLLECTING,
        SEEDING,
        DONE;
    }
}

File: src/main/scala/example/bootstrap/BootstrapServer.scala
package example.bootstrap


import java.net.InetAddress
import java.util.UUID

import com.google.common.collect.ImmutableSet
import example.bootstrap.networking.{Message, NetAddress}
import se.sics.kompics.Start
import se.sics.kompics.network.netty.NettyNetwork
import se.sics.kompics.sl.{ComponentDefinition, _}
import se.sics.kompics.timer.{CancelPeriodicTimeout, SchedulePeriodicTimeout, Timer}

import scala.collection.JavaConverters._
import scala.collection.mutable


object State extends Enumeration {
    type State = Value
    val Collecting, Seeding, Done = Value
}
class BootstrapServer(init: Init[BootstrapServer]) extends ComponentDefinition {
    import example.bootstrap.State.State


    val boot = provides[Bootstrapping]
    var network = requires[NettyNetwork]
    var timer = requires[Timer]

//    val self: String = cfg.getValue("sh.address")
    val self = new NetAddress(InetAddress.getByName("http://localhost"), 8082)

    private var timeoutId: Option[UUID] = None
    private var active:mutable.Set[NetAddress] = collection.mutable.Set[NetAddress]()
    private var ready:mutable.Set[NetAddress] = collection.mutable.Set[NetAddress]()
    private var initialAssignment = None : Option[NodeAssignment]
    private var state:State = State.Collecting


    ctrl uponEvent {
        case _: Start => handle {
            val timeout: Long = 2L // TODO
            val spt = new SchedulePeriodicTimeout(timeout, timeout)
            spt.setTimeoutEvent(new BootstrapTimeout(spt))
            trigger(spt -> timer)
            timeoutId = Some(spt.getTimeoutEvent.getTimeoutId)
            active = active + self
        }
    }

    timer uponEvent {
        case BootstrapTimeout(_) => handle {
            import example.bootstrap.State._

            state match {
                case Collecting =>
                    if (active.size >= 3) {
                        // TODO bootThreshold
                        bootUp()
                    }
                case Seeding =>
                    if (ready.size >= 3) {
                        // TODO bootThreshold
                        trigger(Booted(initialAssignment.get) -> boot)
                        state = Done
                    }
                case Done =>
                    suicide()
            }
        }
    }

    network uponEvent  {
        case context  @ Message(_, from, CheckIn) => handle {
        }

    }

    private def bootUp(): Unit = {
        state = State.Seeding
        trigger(GetInitialAssignments(collection.immutable.Set[NetAddress](active.toSeq: _*)) -> boot)
    }

    override def tearDown(): Unit = {
        trigger(new CancelPeriodicTimeout(timeoutId.get) -> timer)
    }

}
